### DISCUSSION TITLE
Avalonia Support

### DISCUSSION THREAD

**Discussion by flurin137** [canton7/Stylet#342] [Ideas]
Hi together

Would there be a chance that in some point in future [AvaloniaUI](http://avaloniaui.net) would be supported as well? It seems conceptually very similar to WPF but offers cross platform functionality and a out of the box "modern" look.
I find the default XAML toolkit it comes with ([ReactiveUI](https://www.reactiveui.net)) quite cumbersome to use and really love the simplicity and elegance of Stylet.

Regards
Flurin

**Comment by canton7** [comment]
I considered it (I've got a couple of projects using Avalonia), but Avalonia already does a lot of what Stylet does (particularly around the VM-first approach, and providing a way to bind events to VM methods), so there's not a lot for Stylet to *do*. In my experience you can quite happily skip out the ReactiveUI stuff and not really lose anything. Sure things like the `IWindowManager` interface could probably come across -- I haven't looked into that too much.

I'm mainly just waiting to see what MAUI is going to be like, as that's making big promises about being the future of cross-platform .NET UIs.

**Reply by flurin137** [reply]
Thanks for the reply.
Simple binding of events to methods was the particular thing I was trying to achieve in Avalonia but haven't found an easy way to do it. (the way I saw it achieved is using Behaviors which is IMO quite some overhead) 
The thing I miss most is the `Refresh()` method on `PropertyChangedBase`.
I too am really looking forward to see how MAUI will be, it could be quite the game changer but i feat that it might heavily lean towards mobile and made desktop development inconvenient.

**Reply by canton7** [reply]
You can use `{Binding MethodName}` to bind directly to a method, [see here](https://docs.avaloniaui.net/docs/data-binding/binding-to-commands#binding-to-methods).

`Refresh` just raises a `PropertyChanged` event with a propertyName of `string.Empty`, so it's easy to do this using your own `PropertyChanged` base class.

**Reply by flurin137** [reply]
Okay, does the binding of events to VM Methods also work in the same way? 
I knew about the binding of Commands but was not aware of event binding.  The only thing I found is this: [https://github.com/AvaloniaUI/Avalonia/discussions/6171](https://github.com/AvaloniaUI/Avalonia/discussions/6171) which is quite some overhead to `{Bindig MethodName}`.
Good to know that giving `string.Empty` to `PropertyChanged` works as full refresh 👍.

**Reply by canton7** [reply]
Not sure on that one: you'll have to test it. Stylet makes that work by returning a delegate from a MarkupExtension to a RoutedEvent, which is something that's pretty niche: it doesn't work in older WPF or Xamarin.Forms, and I doubt it would work in Avalonia. If Avalonia doesn't support binding events to methods, it's probably going to be easiest, or indeed necessary, to implement this directly in Avalonia rather than in a third party framework.

**Reply by WillBellJr** [reply]
I created a quick MAUI project to get familiar - the application window kept opening up nearly full screen.

Looked into how to change / resize the (Windows) application window (something a baby's teddy bear can easily do in WPF by accidently falling onto the keyboard!) - it was so complex and out of the way, I deleted the project and reopened one of my Stylet projects just to make myself feel better!

I can appreciate the need for code to allow for all the other platforms, but that Windows MAUI code to change the window size, had me feeling like I was back in the days of **Charles Petzold** - I'm good, I can wait a tad longer...

**Comment by alexhelms** [comment]
Adding to this discussion, @canton7 you mention that Avalonia already does VM-first but I'm struggling to see how it does that. I'm probably missing something, but how do you do that in Avalonia out of the box?

**Reply by canton7** [reply]
Through the ViewLocator, see e.g. [here](https://docs.avaloniaui.net/tutorials/todo-list-app/locating-views). You can bind a VM to a UI container element's Content property, and Avalonia will call into the ViewLocator to find and instantiate the correct view.

**Reply by alexhelms** [reply]
Duh, thank for pointing that out, totally did not put that together. As you mentioned in #348, Conductor/Screen are the most agnostic parts. I supposed I'll experiment with porting them to something that is Avalonia compatible. I'm a big fan of the Conductor/Screen architecture.

**Reply by canton7** [reply]
Yeah, I'm playing around with the idea of rewriting them to sit on top of the mvvm community toolkit, maybe as part of a separate package... I'm not sure where if anywhere that train of thought might lead though!

Screen and the Conductors are pretty independent of everything else, so porting then should be pretty straightforward.

**Reply by alexhelms** [reply]
I've made a proof of concept of Conductors and Screens working with Avalonia, have a look in this repo: https://github.com/alexhelms/StyletAvalonia

I started with copy and pasting parts of Stylet into a clean Avalonia project but in retrospec I probably should have pulled the Stylet repo and deleted files. I'm not sure if you've done similar work with porting Stylet to Avalonia but I found it to be fairly compatible. There are a couple things unsupported that I noted in the readme on my repo but it seems that they can all be worked around eventually.

**Reply by canton7** [reply]
Very interesting, thank you! I'll go through the properly when I get the chance.

Would you mind holding off publishing anything for now? I'm trying to work out whether it's worth e.g. bumping Stylet to version 2 and switching to the community toolkit, then using that as the base for ports to other platforms. That would also let us get rid of ValidatingModelBase for instance in favour of ObservableValidator. I'd also like to consider whether to get rid of the MessageBoxView/ViewModel, and improve the story around multiple assemblies. I'm keen that we're as consistent as possible across platforms here.

**Reply by alexhelms** [reply]
Of course, I was only planning on creating that proof of concept repo and nothing past that.

**Reply by WillBellJr** [reply]
> Yeah, I'm playing around with the idea of rewriting them to sit on top of the mvvm community toolkit, maybe as part of a separate package... I'm not sure where if anywhere that train of thought might lead though!
> 
> Screen and the Conductors are pretty independent of everything else, so porting then should be pretty straightforward.

PLEASE do Canton - after playing around w/Community Toolkit v8, I was literally going to pen a message for you to support CT8 - I'm so glad you're considering it!

I absolutely LOVE Stylet, and I've been so successful with my projects using it - I love how w/CT8 all that boilerplate code is gone, but I'm so used to Stylet's _ViewModel First_ paradigm.

I literally scrapped the new CT8 project project I started, and went back to Stylet since CT8 also felt like View first to me??! (I missed my Conductors!)

I also saw myself, w/CT8, wanting to write code in the XAML code behind file where as with Stylet I NEVER had any reason to do so!
I also LOVE Stylet's IoC setup w/the Bootstrapper, and the Actions compared to CT8. (But LAWD all the typing that CT8 saves doh..)

So, again PLEASE if you can, I'd love the code saving of CT8 w/the MVVM simplicity of Stylet.

I'm going to embarrass myself here, but out of desperation, I tried combining Stylet into a CT8 project (they did say "A la carte"!), but (of course) I ran into issues w/the two sets of INotifyPropertyChange clashing between the frameworks...

But yeah, it shows that I WANTED Stylet w/the Code savings of CT8!

**Comment by pokorpa** [comment]
If anyone is still looking for Avalonia solution, I have updated @alexhelms fork to Avalonia 11 and it works (for me).
Source is available https://gitlab.com/pokorpa/stiletto.avalonia
